Fix OpenBSD/arm64 support

dustanddreams opened the pull request:
The changes in this PR allow literals to be emitted in rodata section, rather than in text section.

This is intended to support platforms where the text section is not readable (one such platform being OpenBSD/arm64 7.3).

~While switching to .rodata for literals has no visible performance impact, jump tables require more instructions in order to compute the branch address (the jump tables have to contain pc-relative displacements in order to avoid requiring relocations at binary load time), this change of behaviour is only enabled under OpenBSD.~

Poke @avsm @kayceesrk

dustanddreams commented:
Could someone with enough rights add the "no-change-entry-needed" label to appease the hygiene check? There is no user-visible change in this PR and it does not warrant a Changes entry IMHO.

avsm commented:
Thanks @dustanddreams, I'm running the test suite on this PR on OpenBSD-current/arm64 (on an M1) with no problems so far. 

I think we'll need to remove the hardcoded check for OpenBSD and check for non-readable text segments in the configure scripts.  I got curious and went looking for other platforms where executable-only memory has been implemented, and ended up in a rollercoaster ride of finding that [Android 10 had XOM](https://source.android.com/docs/security/test/execute-only-memory), but it got pulled in Android 11 due to some [speculative process vulnerabilities](https://developer.arm.com/Arm%20Security%20Center/Speculative%20Processor%20Vulnerability#Resources).  There's a description of OpenBSD's xonly [here](https://marc.info/?l=openbsd-tech&m=167374666324119&w=2) and [here](https://www.openbsd.org/papers/csw2023.pdf).

ghost commented:
A problem with checking for text section readability at `configure` time, is that support may be partial on a particular platform (e.g. the `OpenBSD/sparc64` platform will eventually have unreadable text on `sun4u` machines but not on `sun4v` - not a problem for Ocaml since there isn't native code generation for sparc anymore, but you get the idea - configure may run on a less strict machine than the one you intend to run your binaries on).

I think it is better to have that knowledge computed at runtime in `ocamlc`.

avsm commented:
It does need a Changes entry under "code generation", but that can wait until the PR is approved and ready for merging.

I'm seeing some fail tests around dynlinking:

```
List of failed tests:
    tests/lib-dynlink-native/'main.ml' with 1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.2.2.1.1.1.1.1.1.1.1.1.1 (run) Process 1712 got signal 11(Segmentation fault), core dumped
    tests/lib-marshal/'intext.ml' with 1 (native) Process 13448 got signal 11(Segmentation fault), core dumped
    tests/lib-marshal/'intext_par.ml' with 1 (native) Process 32332 got signal 11(Segmentation fault), core dumped
    tests/lib-dynlink-private/'test.ml' with 1.2.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1 (run) Process 49832 got signal 11(Segmentation fault), core dumped
    tests/runtime-objects/'Tests.ml' with 1 (native) Process 59618 got signal 11(Segmentation fault), core dumped
```

ghost commented:
Turns out the dynamic linking code relies upon text being readable, for `caml_digest_of_code_fragment` to run. There is no way this can work on a platform with non-readable text sections.

So maybe the initial approach of linking with `-Wl,--no-execute-only` will be preferrable to cover this situation.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -327,12 +332,16 @@ let float_literal env fl =
   if env.float_literals <> [] then begin
     if macosx then
-      `	.section	__TEXT,__literal8,8byte_literals\n`;
+      `	.section	__TEXT,__literal8,8byte_literals\n`
+    else if not text_section_is_readable then
+      `	.section .rodata\n`;
```
I'm in favor of putting literals in read-only data sections for all ports, incl. Linux.  It should improve code cache utilization a bit.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -1004,25 +1013,33 @@ let emit_instr env i =
-            `	.word	{emit_label jumptbl.(j)} - {emit_label lbltbl}\n`
-        done
-*)
+        if text_section_is_readable then begin
+            let lbltbl = new_label() in
+            `	adr	{emit_reg reg_tmp1}, {emit_label lbltbl}\n`;
+            `	ldr	{emit_reg reg_tmp1}, [{emit_reg reg_tmp1}, {emit_reg i.arg.(0)}, lsl #3]\n`;
+            `	br	{emit_reg reg_tmp1}\n`;
+            `{emit_label lbltbl}:`;
+            for j = 0 to Array.length jumptbl - 1 do
+                `	.quad	{emit_label jumptbl.(j)}\n`
+            done
+        end else begin
+            let lbltbl = new_label() in
+            `	adrp	{emit_reg reg_tmp1}, {emit_label lbltbl}\n`;
+            `	add	{emit_reg reg_tmp1}, {emit_reg reg_tmp1}, #:lo12:{emit_label lbltbl}\n`;
+            `	ldr	{emit_reg reg_tmp1}, [{emit_reg reg_tmp1}, {emit_reg i.arg.(0)}, lsl #3]\n`;
+            let lblbase = new_label() in
+            `{emit_label lblbase}:`;
+            `	adr	{emit_reg reg_tmp2}, .\n`;
+            `	add	{emit_reg reg_tmp1}, {emit_reg reg_tmp1}, {emit_reg reg_tmp2}\n`;
+            `	br	{emit_reg reg_tmp1}\n`;
+            `	.section .rodata\n`;
+            `	.align	3\n`;
+            `{emit_label lbltbl}:`;
+            for j = 0 to Array.length jumptbl - 1 do
+                `	.quad	{emit_label jumptbl.(j)} - {emit_label lblbase}\n`
+            done;
+            `	.previous\n`
+        end
```
You replaced the original code by two new implementations, one that reads from the text section and the other that does not.  But the original code does not read from the text section either!  And is more efficient than either of your new codes.  What's going on here?

dustanddreams replied:
Sorry about that, that's a complete brain fart on my side. I intended to keep the existing code for readable text sections but messed up - and indeed, as you wrote, the original code doesn't need to read from the text section so no change is required there. I'll revert that chunk.

xavierleroy reviewed (commented):
I need to understand why you changed the compilation of `Lswitch`.  See below.

Commit:
Emit literals in rodata section

Floating-point literals are now explicitly put in .rodata rather than
in the current text section, except on MacOS X.

Commit:
Disable execute-only under OpenBSD

The dynlink code (especially caml_digest_of_code_fragment) requires text
sections to be readable in order to perform its work.

avsm commented on configure.ac:
```diff
@@ -940,7 +940,9 @@ AS_CASE([$cc_basename,$host],
   [gcc*,powerpc-*-linux*],
-    [oc_ldflags="-mbss-plt"])
+    [oc_ldflags="-mbss-plt"],
+  [*,*-*-openbsd*],
```
Shouldn't this be just arm64?  The current OCaml trunk works fine for me on OpenBSD-current/amd64.

dustanddreams replied:
Even the dynlink part? If so, the test can be narrowed to arm64.

avsm replied:
Sorry, I tell a lie -- I had natdynlink disabled for other reasons in my checkout.  A completely clean trunk build on amd64 failed the natdynlink tests as well,  and the remaining tests passing on amd64 makes sense since that architecture already places literals in .rodata.

I've kicked off a testsuite run with the latest version of your patch on amd64 and arm64 again.

avsm commented:
It's a pity we don't have a way to signal to the code generator that we need dynlink support in the binary output.  I barely ever use Dynlink and would prefer exec-only rodata as a default, but the OpenBSD package for OCaml can't go around disabling dynlink unless absolutely necessary.

avsm commented on Changes:
```diff
@@ -12,6 +12,9 @@ Working version
 ### Code generation and optimizations:
 
+- #12221: On arm64, emit floating-point literals in the rodata section
+  instead that in the text section, on all platforms but MacOS X.
```
```suggestion
  instead that in the text section, on all platforms but macOS.
```

Just a nit; the downstream name of the OS has changed in recent years.

Commit:
fixup! Emit literals in rodata section

Co-authored-by: Anil Madhavapeddy <anil@recoil.org>

xavierleroy commented:
> It's a pity we don't have a way to signal to the code generator that we need dynlink support in the binary output.

Dynamic linking is not the problem here.  It's marshaling of function closures that triggers the hashing of code blocks, both the statically-linked code from the main program and the dynamically-linked code from plugins.

Commit:
fixup! Disable execute-only under OpenBSD

avsm commented:
This now passes with the mkdll changes to make it no-execute on OpenBSD/amd64, but fails on OpenBSD/arm64 in one test (tests/lib-dynlink-private/test.ml) with a segfault. The other dynlink tests that formerly fail now pass though.

(Apologies for not being able to dig into the core dump more to find out what's wrong -- I'm only keeping half an eye on this while writing lecture notes to a tight deadline!)

ghost commented:
> This now passes with the mkdll changes to make it no-execute on OpenBSD/amd64, but fails on OpenBSD/arm64 in one test (tests/lib-dynlink-private/test.ml) with a segfault. The other dynlink tests that formerly fail now pass though.

That's similar to what I am seeing. It looks like a stack gets corrupted, I have not found the root cause yet.

kayceesrk commented:
@dustanddreams have you had a chance to make progress on finding the root cause?

ghost commented:
> @dustanddreams have you had a chance to make progress on finding the root cause?

No progress so far.
